lib/srv/db/server.go

```diff
@@ -223,8 +223,11 @@ func (s *Server) initHeartbeat(ctx context.Context, server types.DatabaseServer)
 
 func (s *Server) getServerInfoFunc(server types.DatabaseServer) func() (services.Resource, error) {
 	return func() (services.Resource, error) {
-		s.mu.Lock()
-		defer s.mu.Unlock()
+		// Make sure to return a new object, because it gets cached by
+		// heartbeat and will always compare as equal otherwise.
```

Comment:
Following up on your comment - I believe heartbeat's code uses `CompareServers` to compare by value and not by reference but cloning is still required to avoid races.

Reply:
@a-palchikov Yes, it compares by value, but we'd be comparing the same object to itself every time if we didn't return a new one here. And yeah, data races too of course.